% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/makeBins.R
\name{fixedWidthBins}
\alias{fixedWidthBins}
\title{Make fixed-width bins}
\usage{
fixedWidthBins(bamfile = NULL, assembly = NULL, chrom.lengths = NULL,
  chromosome.format, binsizes = 1e+06, chromosomes = NULL)
}
\arguments{
\item{bamfile}{A BAM file from which the header is read to determine the chromosome lengths. If a \code{bamfile} is specified, option \code{assembly} is ignored.}

\item{assembly}{An assembly from which the chromosome lengths are determined. Please see \code{\link[GenomeInfoDb]{fetchExtendedChromInfoFromUCSC}} for available assemblies. This option is ignored if \code{bamfile} is specified. Alternatively a data.frame generated by \code{\link[GenomeInfoDb]{fetchExtendedChromInfoFromUCSC}}.}

\item{chrom.lengths}{A named character vector with chromosome lengths. Names correspond to chromosomes.}

\item{chromosome.format}{A character specifying the format of the chromosomes if \code{assembly} is specified. Either 'NCBI' for (1,2,3 ...) or 'UCSC' for (chr1,chr2,chr3 ...). If a \code{bamfile} or \code{chrom.lengths} is supplied, the format will be chosen automatically.}

\item{binsizes}{A vector of bin sizes in base pairs.}

\item{chromosomes}{A subset of chromosomes for which the bins are generated.}
}
\value{
A \code{list()} of \code{\link{GRanges}} objects with fixed-width bins.
}
\description{
Make fixed-width bins based on given bin size.
}
\examples{
## Make fixed-width bins of size 500kb and 1Mb
data(rn4_chrominfo)
chrom.lengths <- rn4_chrominfo$length
names(chrom.lengths) <- rn4_chrominfo$chromosome
bins <- fixedWidthBins(chrom.lengths=chrom.lengths, binsizes=c(5e5,1e6))
bins

## Make bins using NCBI server (requires internet connection)
# bins <- fixedWidthBins(assembly='mm10', chromosome.format='NCBI', binsizes=c(5e5,1e6))

}
\author{
Aaron Taudt
}

