\name{make.ucsc.file.univariate}
\alias{make.ucsc.file.univariate}
\title{
Convert univariate HMM models for upload to the UCSC genome browser
}
\description{
Convert the unmodified/modified calls from univariate HMM models to a BED file which is ready to upload and display in the UCSC Genome Browser.
}
\usage{
make.ucsc.file.univariate(modellist, namevec, filename="univariate-calls-ucsc", threshold=0.5)
}
\arguments{
\item{modellist}{
A list of models generated by \code{\link{univariate.from.binned.data}}, e.g. 'list(model1,model2,...)'.
}
\item{namevec}{
A vector of strings which will be used to label the tracks that are created from the 'modellist'.
}
\item{file}{
Filename of the output file.
}
\item{threshold}{
Threshold for calling a bin modified or unmodified. If the posterior for state 'modified' is greater than the threshold, then this bin will be called modified.
}
}

\details{
This function expects a list of models, which were generated by the function \code{\link{univariate.from.binned.data}}. The posteriors of each model will be converted to unmodified/modified calls and written to a specified file in BED format. This file can then be readily imported into the UCSC Genome Browser.
}
\keyword{ucsc}
\keyword{browser}
\references{}
\author{Aaron Taudt}
\note{}
\seealso{}
\examples{
# Load example datasets with genome coordinates and number of aligned reads for each bin
data(binned.data.example.1)
data(binned.data.example.2)
# Do the univariate HMMs
model1 = univariate.from.binned.data(binned.data.example.1, eps=1, max.time=60)
model2 = univariate.from.binned.data(binned.data.example.2, eps=1, max.time=60)
# Convert the models to BED files
make.ucsc.file.univariate(list(model1,model2),c("mark1","mark2"),file="upload_this_to_the_ucsc")
}
