\name{make.ucsc.file.multivariate}
\alias{make.ucsc.file.multivariate}
\title{
Convert a multivariate HMM model for upload to the UCSC genome browser
}
\description{
Convert the combinatorial states from a multivariate HMM models to a BED file which is ready to upload and display in the UCSC Genome Browser.
}
\usage{
make.ucsc.file.multivariate(model, numstates=NULL, file="multivariate-states-ucsc")
}
\arguments{
\item{modellist}{
A list of models generated by \code{\link{multivariate.from.univariate.results}}, e.g. 'list(model1,model2,...)'.
}
\item{numstates}{
The number of combinatorial states which shall be exported. If unset or greater than the number of combinatorial states possible, the maximum number of combinatorial states will be used (all states that appear at least once).
}
\item{file}{
Filename of the output file.
}
}

\details{
This function expects a list of models, which were generated by the function \code{\link{multivariate.from.univariate.results}}. The posteriors of the model will be converted to the decimal combinatorial states, which will be written to a specified file in BED format. This file can then be readily imported and displayed in the UCSC Genome Browser.
}
\keyword{ucsc}
\keyword{browser}
\references{}
\author{Aaron Taudt}
\note{}
\seealso{}
\examples{
# Load example datasets with genome coordinates and number of aligned reads for each bin
data(binned.data.example.1)
data(binned.data.example.2)
# Do the univariate HMMs
model1 = univariate.from.binned.data(binned.data.example.1, eps=1, max.time=60)
model2 = univariate.from.binned.data(binned.data.example.2, eps=1, max.time=60)
# Do the multivariate HMM
multimodel = multivariate.from.univariate.results(list(model1,model2), eps=1, max.time=180)
# Convert the models to BED files
make.ucsc.file.multivariate(multimodel,file="upload_this_to_the_ucsc")
}
