%\VignetteIndexEntry{The chromstaR users guide}
%\VignetteEngine{knitr::knitr}
\documentclass[11pt]{article}\usepackage[]{graphicx}\usepackage[]{color}
%% maxwidth is the original width if it is less than linewidth
%% otherwise use linewidth (to make sure the graphics do not exceed the margin)
\makeatletter
\def\maxwidth{ %
  \ifdim\Gin@nat@width>\linewidth
    \linewidth
  \else
    \Gin@nat@width
  \fi
}
\makeatother

\definecolor{fgcolor}{rgb}{0.345, 0.345, 0.345}
\newcommand{\hlnum}[1]{\textcolor[rgb]{0.686,0.059,0.569}{#1}}%
\newcommand{\hlstr}[1]{\textcolor[rgb]{0.192,0.494,0.8}{#1}}%
\newcommand{\hlcom}[1]{\textcolor[rgb]{0.678,0.584,0.686}{\textit{#1}}}%
\newcommand{\hlopt}[1]{\textcolor[rgb]{0,0,0}{#1}}%
\newcommand{\hlstd}[1]{\textcolor[rgb]{0.345,0.345,0.345}{#1}}%
\newcommand{\hlkwa}[1]{\textcolor[rgb]{0.161,0.373,0.58}{\textbf{#1}}}%
\newcommand{\hlkwb}[1]{\textcolor[rgb]{0.69,0.353,0.396}{#1}}%
\newcommand{\hlkwc}[1]{\textcolor[rgb]{0.333,0.667,0.333}{#1}}%
\newcommand{\hlkwd}[1]{\textcolor[rgb]{0.737,0.353,0.396}{\textbf{#1}}}%

\usepackage{framed}
\makeatletter
\newenvironment{kframe}{%
 \def\at@end@of@kframe{}%
 \ifinner\ifhmode%
  \def\at@end@of@kframe{\end{minipage}}%
  \begin{minipage}{\columnwidth}%
 \fi\fi%
 \def\FrameCommand##1{\hskip\@totalleftmargin \hskip-\fboxsep
 \colorbox{shadecolor}{##1}\hskip-\fboxsep
     % There is no \\@totalrightmargin, so:
     \hskip-\linewidth \hskip-\@totalleftmargin \hskip\columnwidth}%
 \MakeFramed {\advance\hsize-\width
   \@totalleftmargin\z@ \linewidth\hsize
   \@setminipage}}%
 {\par\unskip\endMakeFramed%
 \at@end@of@kframe}
\makeatother

\definecolor{shadecolor}{rgb}{.97, .97, .97}
\definecolor{messagecolor}{rgb}{0, 0, 0}
\definecolor{warningcolor}{rgb}{1, 0, 1}
\definecolor{errorcolor}{rgb}{1, 0, 0}
\newenvironment{knitrout}{}{} % an empty environment to be redefined in TeX

\usepackage{alltt}
\usepackage{hyperref}
\usepackage{url}
\usepackage[authoryear,round]{natbib}
\bibliographystyle{plainnat}

\newcommand{\scscst}{\scriptscriptstyle}
\newcommand{\scst}{\scriptstyle}

\newcommand{\Rfunction}[1]{{\texttt{#1}}}
\newcommand{\Robject}[1]{{\texttt{#1}}}
\newcommand{\Rpackage}[1]{{\textit{#1}}}

\author{Aaron Taudt\footnote{aaron.taudt@gmail.com}}
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\begin{document}
\title{The chromstaR user's guide}

\maketitle

\tableofcontents
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Introduction}

ChIP-seq has become the standard technique for assessing the genome-wide chromatin state of DNA. \Rpackage{chromstaR} provides functions for the joint analysis of multiple ChIP-seq samples. It allows peak calling for transcription factor binding and histone modifications with a narrow (e.g. H3K4me3, H3K27ac,~...) or broad (e.g. H3K36me3, H3K27me3,~...) profile. All analysis can be performed on each sample individually (=univariate), or in a joint analysis considering all samples simultaneously (=multivariate).



\section{Outline of workflow}

Every analysis with the \Rpackage{chromstaR} package starts from aligned reads in either BAM or BED format. In the first step, the genome is partitioned into non-overlapping, equally sized bins and the reads that fall into each bin are counted. These read counts serve as the basis for both the univariate and the multivariate peak- and broad-region calling. Univariate peak calling is done by fitting a three-state Hidden Markov Model to the binned read counts. Multivariate peak calling for $\mathcal{S}$ samples is done by fitting a $2^\mathcal{S}$-state Hidden Markov Model to all binned read counts.

%' \section{Univariate analysis}
%' 
%' \subsection{\label{sec:narrow}Task 1: Peak calling for a narrow histone modification}
%' 
%' Examples of histone modifications with a narrow profile are H3K4me3, H3K9ac and H3K27ac in most human tissues. For such peak-like modifications, the bin size should be set to a value between 200bp and 1000bp.
%' 
%' \begin{scriptsize}
%' <<univariate_narrow_library, results='hide', message=FALSE, eval=TRUE>>=
%' library(chromstaR)
%' @
%' 
%' <<univariate_narrow_binning, results='show', message=FALSE, eval=TRUE>>=
%' ## === Step 1: Binning ===
%' # Get an example BED file
%' bedfile <- system.file("extdata","euratrans","liver-H3K4me3-BN-male-bio2-tech1.bed.gz",
%'                         package="chromstaRData")
%' # Get the chromosome lengths (see ?GenomeInfoDb::fetchExtendedChromInfoFromUCSC)
%' # This is only necessary for BED files. BAM files are handled automatically.
%' chrom.lengths <- data(rn4_chrominfo)
%' # We use bin size 1000bp and chromosome 12 to keep the example quick
%' binned.data <- binReads(bedfile, format='bed',assembly=rn4_chrominfo, binsizes=1000,
%'                         chromosomes='chr12')
%' print(binned.data)
%' @
%' 
%' <<univariate_narrow_peak_calling, results='markup', eval=TRUE>>=
%' ## === Step 2: Peak calling ===
%' # We restrict the peak calling to 60 seconds to keep this example quick.
%' model <- callPeaksUnivariate(binned.data, ID='H3K4me3', max.time=60, verbosity=0)
%' @
%' 
%' <<univariate_narrow_plotting, fig.width=6, fig.height=4>>=
%' ## === Step 3: Checking the fit ===
%' # For a narrow modification, the fit should look something like this,
%' # with the 'modified'-component near the bottom of the figure
%' plot(model)
%' @
%' 
%' <<univariate_narrow_export, results='hide', message=FALSE, eval=FALSE>>=
%' ## === Step 4: Export to genome browser ===
%' # We can export peak calls and binned read counts with
%' exportUnivariates(list(model), filename='your-peak-file', what='peaks')
%' exportUnivariates(list(model), filename='your-read-file', what='counts')
%' @
%' \end{scriptsize}
%' 
%' \subsection{\label{sec:broad}Task 2: Peak calling for a broad histone modification}
%' 
%' Examples of histone modifications with a broad profile are H3K9me3, H3K27me3, H3K36me3, H4K20me1 in most human tissues. These modifications usually cover broad domains of the genome, and the enrichment is best captured with a bin size between 500bp and 2000bp.
%' 
%' \begin{scriptsize}
%' <<univariate_broad_library, results='hide', message=FALSE, eval=TRUE>>=
%' library(chromstaR)
%' @
%' 
%' <<univariate_broad_binning, results='hide', message=FALSE, eval=TRUE>>=
%' ## === Step 1: Binning ===
%' # Get an example BED file
%' bedfile <- system.file("extdata","euratrans","liver-H3K27me3-BN-male-bio1-tech1.bed.gz",
%'                         package="chromstaRData")
%' # Get the chromosome lengths (see ?GenomeInfoDb::fetchExtendedChromInfoFromUCSC)
%' # This is only necessary for BED files. BAM files are handled automatically.
%' chrom.lengths <- data(rn4_chrominfo)
%' # We use bin size 1000bp and chromosome 12 to keep the example quick
%' binned.data <- binReads(bedfile, format='bed',assembly=rn4_chrominfo, binsizes=1000,
%'                         chromosomes='chr12')
%' print(binned.data)
%' @
%' 
%' <<univariate_broad_peak_calling, results='markup', eval=TRUE>>=
%' ## === Step 2: Peak calling ===
%' # We restrict the peak calling to 60 seconds to keep this example quick.
%' model <- callPeaksUnivariate(binned.data, ID='H3K27me3', max.time=60, verbosity=0)
%' @
%' 
%' <<univariate_broad_plotting, fig.width=6, fig.height=4>>=
%' ## === Step 3: Checking the fit ===
%' # For a broad modification, the fit should look something like this,
%' # with a 'modified'-component that fits the thick tail of the distribution.
%' plot(model)
%' @
%' 
%' <<univariate_broad_export, results='hide', message=FALSE, eval=FALSE>>=
%' ## === Step 4: Export to genome browser ===
%' # We can export peak calls and binned read counts with
%' exportUnivariates(list(model), filename='your-peak-file', what='peaks')
%' exportUnivariates(list(model), filename='your-read-file', what='counts')
%' @
%' 
%' <<univariate_broad_H4K20me1, echo=TRUE, results='hide', message=FALSE, fig.width=6, fig.height=4>>=
%' ## === Step 1-3: Another example for mark H4K20me1 ===
%' bedfile <- system.file("extdata","euratrans","liver-H4K20me1-BN-male-bio1-tech1.bed.gz",
%'                        package="chromstaRData")
%' data(rn4_chrominfo)
%' binned.data <- binReads(bedfile, format='bed',assembly=rn4_chrominfo, binsizes=1000,
%'                         chromosomes='chr12')
%' model <- callPeaksUnivariate(binned.data, ID='H4K20me1', max.time=60, verbosity=0)
%' plot(model)
%' @
%' 
%' \end{scriptsize}
%' 
%' \subsection{Task 3: Peak calling for ATAC-seq, DNase-seq, FAIRE-seq, ...}
%' 
%' Peak calling for ATAC-seq and DNase-seq is similar to the peak calling of a narrow histone modification (see section~\ref{sec:narrow}). FAIRE-seq experiments seem to exhibit a broad profile with our model, so the procedure is similar to the domain calling of a broad histone modification (see section~\ref{sec:broad}).
%' 
%' \section{Multivariate analysis}
%' \subsection{Task 1: Integrating multiple replicates}
%' 
%' \Rpackage{chromstaR} can be used to call peaks with multiple replicates, without the need of prior merging. The underlying statistical model integrates information from all replicates to identify common peaks. It is, however, important to note that replicates with poor quality can affect the joint peak calling negatively. It is therefore recommended to first check the replicate quality and discard poor-quality replicates. The necessary steps for peak calling for an example ChIP-seq experiment with 4 replicates are detailed below.
%' 
%' \begin{scriptsize}
%' <<univariate_replicate_library, results='hide', message=FALSE, eval=TRUE>>=
%' library(chromstaR)
%' @
%' 
%' <<multivariate_replicate_binning, results='show', message=FALSE, eval=TRUE>>=
%' ## === Step 1: Binning ===
%' # Let's get some example data with 3 replicates
%' file.path <- system.file("extdata","euratrans", package='chromstaRData')
%' bedfiles.good <- list.files(file.path, pattern="liver.*H3K27me3", full.names=TRUE)[1:3]
%' # We fake a replicate with poor quality by taking a different mark entirely
%' bedfiles.poor <- list.files(file.path, pattern="liver.*H4K20me1", full.names=TRUE)[1]
%' bedfiles <- c(bedfiles.good, bedfiles.poor)
%' # This is only necessary for BED files. BAM files are handled automatically.
%' data(rn4_chrominfo)
%' # We use bin size 1000bp and chromosome 12 to keep the example quick
%' binned.data <- list()
%' for (bedfile in bedfiles) {
%'   binned.data[[basename(bedfile)]] <- binReads(bedfile, format='bed', binsize=1000,
%'                                          assembly=rn4_chrominfo, chromosomes='chr12')
%' }
%' @
%' 
%' <<multivariate_replicate_univariate, results='hide', message=FALSE, eval=TRUE>>=
%' ## === Step 2: Univariate peak calling ===
%' # The univariate fit is obtained for each replicate
%' models <- list()
%' for (i1 in 1:length(binned.data)) {
%'   models[[i1]] <- callPeaksUnivariate(binned.data[[i1]], ID=paste0('Rep',i1),
%'                                       max.time=60)
%' }
%' @
%' 
%' <<multivariate_replicate_peak_calling, results='show', message=FALSE, eval=TRUE>>=
%' ## === Step 3: Check replicate correlation ===
%' # We run a multivariate peak calling on all 4 replicates
%' # A warning is issued because replicate 4 is very different from the others
%' multi.model <- callPeaksReplicates(models, max.time=60, eps=1)
%' # Checking the correlation confirms that Rep4 is very different from the others
%' print(multi.model$replicateInfo$correlation)
%' @
%' 
%' <<multivariate_replicate_peak_calling2, results='hide', message=FALSE, eval=TRUE>>=
%' ## === Step 4: Peak calling with replicates ===
%' # We redo the previous step without the "bad" replicate
%' # Also, we force all replicates to agree in their peak calls
%' multi.model <- callPeaksReplicates(models[1:3], force.equal=TRUE, max.time=60)
%' @
%' 
%' <<multivariate_replicate_export, results='hide', message=FALSE, eval=FALSE>>=
%' ## === Step 5: Export to genome browser ===
%' # Finally, we can export the results as BED file
%' exportMultivariate(multi.model, filename='your-peak-file', what='peaks')
%' exportMultivariate(multi.model, filename='your-read-file', what='counts')
%' @
%' \end{scriptsize}
%' 
%' \subsection{Task 2: Detecting differentially modified regions}
%' 
%' \Rpackage{chromstaR} is extremely powerful in detecting differentially modified regions in two or more samples. The following example illustrates this on ChIP-seq data for H4K20me1 in liver and heart (left-ventricle) tissues from rat. With 2 samples we can have $2^2 = 4$ combinatorial states, which can be readily interpreted as '0: all samples unmodified', '1-2: DMR' and '3: all samples modified'. Having several replicates for each sample makes it more complicated, but you get the idea ...
%' 
%' \begin{scriptsize}
%' <<univariate_differential_library, results='hide', message=FALSE, eval=TRUE>>=
%' library(chromstaR)
%' @
%' 
%' <<multivariate_differential_binning, results='hide', message=FALSE, eval=TRUE>>=
%' ## === Step 1: Binning ===
%' # Let's get some example data for heart (lv) and liver with 3 replicates each
%' file.path <- system.file("extdata","euratrans", package='chromstaRData')
%' bedfiles <- list.files(file.path, pattern="H4K20me1", full.names=TRUE)
%' # This is only necessary for BED files. BAM files are handled automatically.
%' data(rn4_chrominfo)
%' # We use bin size 1000bp and chromosome 12 to keep the example quick
%' binned.data <- list()
%' for (bedfile in bedfiles) {
%'   binned.data[[basename(bedfile)]] <- binReads(bedfile, format='bed', binsize=1000,
%'                                          assembly=rn4_chrominfo, chromosomes='chr12')
%' }
%' @
%' 
%' <<multivariate_differential_univariate, results='hide', message=FALSE, eval=TRUE>>=
%' ## === Step 2: Univariate peak calling ===
%' # The univariate fit is obtained for each replicate
%' models <- list()
%' for (i1 in 1:length(binned.data)) {
%'   models[[i1]] <- callPeaksUnivariate(binned.data[[i1]], ID=names(binned.data)[i1],
%'                                       max.time=60)
%' }
%' @
%' 
%' <<multivariate_differential_stateBrewer, results='markup', message=TRUE, eval=TRUE>>=
%' ## === Step 3: Constructing the combinatorial states ===
%' # This step is only necessary if you have replicates for each sample.
%' # To ensure that replicates are treated as such, and not as independent
%' # samples, we have to construct the proper combinatorial states:
%' 
%' # First, we get all the tissues (we could specify them by hand, but we are lazy)
%' IDs <- names(binned.data)
%' tissues <- sapply(strsplit(IDs,'-'),'[[',1)
%' marks <- sapply(strsplit(IDs,'-'),'[[',2)
%' print(tissues)
%' # Second, we obtain the combinatorial states
%' # Look up ?stateBrewer on how to use this function
%' states <- stateBrewer(tissues)
%' print(states)
%' # Third, we construct differential states
%' diff.states <- stateBrewer(tissues, differential.states=TRUE, conditions=tissues,
%'                            tracks2compare=marks)
%' print(diff.states)
%' @
%' 
%' <<multivariate_differential_multivariate, results='hide', message=FALSE, eval=TRUE>>=
%' ## === Step 4: Multivariate peak calling ===
%' multi.model <- callPeaksMultivariate(models, use.states=states, eps=1, max.time=60)
%' @
%' 
%' <<multivariate_differential_export, results='hide', message=FALSE, eval=FALSE>>=
%' ## === Step 5: Export to genome browser ===
%' # Export only differential peaks by excluding the 'common.states'
%' exportMultivariate(multi.model, filename='your-peak-file', what='peaks',
%'                    include.states=diff.states)
%' exportMultivariate(multi.model, filename='your-read-file', what='counts',
%'                    include.states=diff.states)
%' exportMultivariate(multi.model, filename='your-combstates-file', what='combstates',
%'                    include.states=diff.states)
%' @
%' \end{scriptsize}

\subsection{Task 3: Finding combinatorial chromatin states}

Most experimental studies that probe several histone modifications are interested in combinatorial chromatin states. An example of a simple combinatorial state would be [H3K4me3+H3K27me3], which is also frequently called ``bivalent promoter'', due to the simultaneous occurrence of the promoter marking H3K4me3 and the repressive H3K27me3. Finding combinatorial states with \Rpackage{chromstaR} is equivalent to a multivariate peak calling. The following code chunks demonstrate how to find bivalent promoters and do some simple analysis:

\begin{scriptsize}
\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlkwd{library}\hlstd{(chromstaR)}
\end{alltt}
\end{kframe}
\end{knitrout}

\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlcom{## === Step 1: Binning ===}
\hlcom{# Let's get some example data for H3K27me3 and H3K4me3 in rat liver}
\hlstd{file.path} \hlkwb{<-} \hlkwd{system.file}\hlstd{(}\hlstr{"extdata"}\hlstd{,}\hlstr{"euratrans"}\hlstd{,} \hlkwc{package}\hlstd{=}\hlstr{'chromstaRData'}\hlstd{)}
\hlstd{bedfiles} \hlkwb{<-} \hlkwd{list.files}\hlstd{(file.path,} \hlkwc{pattern}\hlstd{=}\hlstr{"liver.*H3K4me3|liver.*H3K27me3"}\hlstd{,}
                       \hlkwc{full.names}\hlstd{=}\hlnum{TRUE}\hlstd{)}
\hlcom{# This is only necessary for BED files. BAM files are handled automatically.}
\hlkwd{data}\hlstd{(rn4_chrominfo)}
\hlcom{# We use bin size 1000bp and chromosome 12 to keep the example quick}
\hlstd{binned.data} \hlkwb{<-} \hlkwd{list}\hlstd{()}
\hlkwa{for} \hlstd{(bedfile} \hlkwa{in} \hlstd{bedfiles) \{}
  \hlstd{binned.data[[}\hlkwd{basename}\hlstd{(bedfile)]]} \hlkwb{<-} \hlkwd{binReads}\hlstd{(bedfile,} \hlkwc{format}\hlstd{=}\hlstr{'bed'}\hlstd{,} \hlkwc{binsize}\hlstd{=}\hlnum{1000}\hlstd{,}
                                         \hlkwc{assembly}\hlstd{=rn4_chrominfo,} \hlkwc{chromosomes}\hlstd{=}\hlstr{'chr12'}\hlstd{)}
\hlstd{\}}
\end{alltt}
\end{kframe}
\end{knitrout}

\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlcom{## === Step 2: Univariate peak calling ===}
\hlcom{# The univariate fit is obtained for each replicate}
\hlstd{models} \hlkwb{<-} \hlkwd{list}\hlstd{()}
\hlkwa{for} \hlstd{(i1} \hlkwa{in} \hlnum{1}\hlopt{:}\hlkwd{length}\hlstd{(binned.data)) \{}
  \hlstd{models[[i1]]} \hlkwb{<-} \hlkwd{callPeaksUnivariate}\hlstd{(binned.data[[i1]],} \hlkwc{ID}\hlstd{=}\hlkwd{names}\hlstd{(binned.data)[i1],}
                                      \hlkwc{max.time}\hlstd{=}\hlnum{60}\hlstd{)}
\hlstd{\}}
\end{alltt}
\end{kframe}
\end{knitrout}

\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlcom{## === Step 3: Constructing the combinatorial states ===}
\hlcom{# This step is only necessary if you have replicates for each sample.}
\hlcom{# To ensure that replicates are treated as such, and not as independent}
\hlcom{# samples, we have to construct the proper combinatorial states:}

\hlcom{# First, we get all the marks (we could specify them by hand, but we are lazy)}
\hlstd{IDs} \hlkwb{<-} \hlkwd{names}\hlstd{(binned.data)}
\hlstd{marks} \hlkwb{<-} \hlkwd{sapply}\hlstd{(}\hlkwd{strsplit}\hlstd{(IDs,}\hlstr{'-'}\hlstd{),}\hlstr{'[['}\hlstd{,}\hlnum{2}\hlstd{)}
\hlkwd{print}\hlstd{(marks)}
\end{alltt}
\begin{verbatim}
## [1] "H3K27me3" "H3K27me3" "H3K27me3" "H3K4me3"  "H3K4me3"  "H3K4me3"
\end{verbatim}
\begin{alltt}
\hlcom{# Second, we obtain the combinatorial states}
\hlcom{# Look up ?stateBrewer on how to use this function}
\hlstd{states} \hlkwb{<-} \hlkwd{stateBrewer}\hlstd{(marks)}
\hlkwd{print}\hlstd{(states)}
\end{alltt}
\begin{verbatim}
##        combination state
## 1                      0
## 2          H3K4me3     7
## 3         H3K27me3    56
## 4 H3K27me3+H3K4me3    63
\end{verbatim}
\end{kframe}
\end{knitrout}

\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlcom{## === Step 4: Multivariate peak calling ===}
\hlstd{multi.model} \hlkwb{<-} \hlkwd{callPeaksMultivariate}\hlstd{(models,} \hlkwc{use.states}\hlstd{=states,} \hlkwc{eps}\hlstd{=}\hlnum{1}\hlstd{,} \hlkwc{max.time}\hlstd{=}\hlnum{60}\hlstd{)}
\end{alltt}
\begin{verbatim}
## HMM: number of states = 4
## HMM: number of bins = 46782
## HMM: maximum number of iterations = none
## HMM: maximum running time = 60 sec
## HMM: epsilon = 1
## HMM: number of experiments = 6
##  Iteration              log(P)             dlog(P)    Time in sec
##          0                -inf                   -              0
## HMM: Precomputing densities ...
##  Iteration              log(P)             dlog(P)    Time in sec
##          0                -inf                   -              0
##          1      -629970.204822                 inf              0
##          2      -628146.574559         1823.630263              0
##          3      -627956.663999          189.910560              0
##          4      -627909.013321           47.650678              0
##          5      -627894.914770           14.098550              0
##          6      -627889.957698            4.957072              0
##          7      -627887.910231            2.047467              0
##          8      -627886.945380            0.964851              0
## HMM: Convergence reached!
## HMM: Recoding posteriors ...
\end{verbatim}
\end{kframe}
\end{knitrout}

\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlcom{## === Step 5: Export to genome browser ===}
\hlcom{# Export combinatorial states}
\hlkwd{exportMultivariate}\hlstd{(multi.model,} \hlkwc{filename}\hlstd{=}\hlstr{'your-file'}\hlstd{,} \hlkwc{what}\hlstd{=}\hlkwd{c}\hlstd{(}\hlstr{'peaks'}\hlstd{,}\hlstr{'counts'}\hlstd{))}
\hlkwd{exportMultivariate}\hlstd{(multi.model,} \hlkwc{filename}\hlstd{=}\hlstr{'your-combstates'}\hlstd{,} \hlkwc{what}\hlstd{=}\hlkwd{c}\hlstd{(}\hlstr{'combstates'}\hlstd{))}
\end{alltt}
\end{kframe}
\end{knitrout}

\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlcom{## === Step 6: Enrichment analysis ===}
\hlcom{# Get coordinates of rat genes}
\hlkwd{library}\hlstd{(biomaRt)}
\hlstd{ensembl} \hlkwb{<-} \hlkwd{useMart}\hlstd{(}\hlstr{'ENSEMBL_MART_ENSEMBL'}\hlstd{,} \hlkwc{host}\hlstd{=}\hlstr{'may2012.archive.ensembl.org'}\hlstd{,}
                    \hlkwc{dataset}\hlstd{=}\hlstr{'rnorvegicus_gene_ensembl'}\hlstd{)}
\hlstd{genes} \hlkwb{<-} \hlkwd{getBM}\hlstd{(}\hlkwc{attributes}\hlstd{=}\hlkwd{c}\hlstd{(}\hlstr{'ensembl_gene_id'}\hlstd{,} \hlstr{'chromosome_name'}\hlstd{,} \hlstr{'start_position'}\hlstd{,}
                            \hlstr{'end_position'}\hlstd{,} \hlstr{'strand'}\hlstd{,} \hlstr{'external_gene_id'}\hlstd{),}
               \hlkwc{mart}\hlstd{=ensembl)}
\hlcom{# Transform to GRanges for easier handling}
\hlstd{genes} \hlkwb{<-} \hlkwd{GRanges}\hlstd{(}\hlkwc{seqnames}\hlstd{=}\hlkwd{paste0}\hlstd{(}\hlstr{'chr'}\hlstd{,genes}\hlopt{$}\hlstd{chrom),}
                 \hlkwc{ranges}\hlstd{=}\hlkwd{IRanges}\hlstd{(}\hlkwc{start}\hlstd{=genes}\hlopt{$}\hlstd{start,} \hlkwc{end}\hlstd{=genes}\hlopt{$}\hlstd{end),}
                 \hlkwc{strand}\hlstd{=genes}\hlopt{$}\hlstd{strand,}
                 \hlkwc{ID}\hlstd{=genes}\hlopt{$}\hlstd{ensembl_gene_id,} \hlkwc{name}\hlstd{=genes}\hlopt{$}\hlstd{external_gene_id)}
\hlkwd{print}\hlstd{(genes)}
\end{alltt}
\begin{verbatim}
## GRanges object with 29516 ranges and 2 metadata columns:
##           seqnames                 ranges strand |                 ID
##              <Rle>              <IRanges>  <Rle> |        <character>
##       [1]    chr13     [1120899, 1121213]      - | ENSRNOG00000043314
##       [2]    chr13     [1192186, 2293551]      - | ENSRNOG00000031539
##       [3]    chr13     [3174383, 3175216]      + | ENSRNOG00000028603
##       [4]    chr13     [4377731, 4379174]      - | ENSRNOG00000030028
##       [5]    chr13     [4866302, 4866586]      - | ENSRNOG00000040235
##       ...      ...                    ...    ... .                ...
##   [29512]     chr6 [134310258, 134310338]      + | ENSRNOG00000035137
##   [29513]     chr9 [  6920889,   6921049]      - | ENSRNOG00000035407
##   [29514]    chr11 [ 40073746,  40073816]      - | ENSRNOG00000043706
##   [29515]     chr2 [233090372, 233090478]      - | ENSRNOG00000035272
##   [29516]     chr6 [ 92917449,  92917541]      + | ENSRNOG00000045521
##                  name
##           <character>
##       [1]   LOC682397
##       [2]   LOC304725
##       [3]            
##       [4]  D3ZPH4_RAT
##       [5]  F1LZC7_RAT
##       ...         ...
##   [29512]    SNORD113
##   [29513]          U1
##   [29514]    SNORD19B
##   [29515]          U6
##   [29516]            
##   -------
##   seqinfo: 22 sequences from an unspecified genome; no seqlengths
\end{verbatim}
\end{kframe}
\end{knitrout}

\end{scriptsize}

\section{\label{sec:faq}FAQ}

\section{Session Info}
\begin{scriptsize}
\begin{knitrout}
\definecolor{shadecolor}{rgb}{0.969, 0.969, 0.969}\color{fgcolor}\begin{kframe}
\begin{alltt}
\hlkwd{sessionInfo}\hlstd{()}
\end{alltt}
\begin{verbatim}
## R Under development (unstable) (2016-02-17 r70182)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 14.04.4 LTS
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=nl_NL.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=nl_NL.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=nl_NL.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=nl_NL.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] biomaRt_2.27.2        chromstaR_0.98.0      chromstaRData_0.99.0 
##  [4] ggplot2_2.1.0         GenomicRanges_1.23.25 GenomeInfoDb_1.7.6   
##  [7] IRanges_2.5.40        S4Vectors_0.9.44      BiocGenerics_0.17.3  
## [10] knitr_1.12.3          devtools_1.10.0      
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.4                 AnnotationDbi_1.33.7       
##  [3] XVector_0.11.7              magrittr_1.5               
##  [5] GenomicAlignments_1.7.20    zlibbioc_1.17.1            
##  [7] BiocParallel_1.5.21         munsell_0.4.3              
##  [9] colorspace_1.2-6            highr_0.5.1                
## [11] stringr_1.0.0               plyr_1.8.3                 
## [13] tools_3.3.0                 SummarizedExperiment_1.1.22
## [15] grid_3.3.0                  Biobase_2.31.3             
## [17] gtable_0.2.0                DBI_0.3.1                  
## [19] digest_0.6.9                reshape2_1.4.1             
## [21] formatR_1.3                 bitops_1.0-6               
## [23] RCurl_1.95-4.8              RSQLite_1.0.0              
## [25] memoise_1.0.0               evaluate_0.8.3             
## [27] stringi_1.0-1               Rsamtools_1.23.6           
## [29] Biostrings_2.39.12          scales_0.4.0               
## [31] XML_3.98-1.4
\end{verbatim}
\begin{alltt}
\hlkwd{warnings}\hlstd{()}
\end{alltt}
\begin{verbatim}
## NULL
\end{verbatim}
\end{kframe}
\end{knitrout}
\end{scriptsize}

\end{document}

